src/controller/CHIPDeviceControllerFactory.cpp

```diff
@@ -317,17 +321,17 @@ void DeviceControllerFactory::Shutdown()
 {
     if (mSystemState != nullptr)
     {
-        mSystemState->Release();
-        chip::Platform::Delete(mSystemState);
+        Platform::Delete(mSystemState);
```

Comment:
Again, this is deleting a refcounted object... importantly without calling `Shutdown` on it.  And `Shutdown` is what does the cleanup of all its members right now.  And it might not have been called yet.  Consider the case of someone doing a factory Init followed by a factory Shutdown with no other calls in between.  We will start off with a zero refcount, never increment it, never decrement it back to 0, never call Shutdown on the state, and then trigger the destructor assert.  In fact the `testFactoryLifecycle` Darwin test does just that.

Reply:
It does not seem valid for me for the System State to outlive the factory. I would prefer the system state have a strict ownership model and that this refcounting go away. The current design is subtle and does not make very much sense.

Reply:
Actually the more straightforward design would be to just move the system state to data members of the factory and totally remove the refcount (or keep it strictly as an error check).

Then the python code wouldn't have to manually and purposefully leak a reference to control the lifecycle.